contracts/test/calldata.spec.ts

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+import { TestCalldataEncryption } from '../typechain-types/contracts/tests';
+import {
+  boxKeyPairFromSecretKey,
+  crypto_box_SECRETKEYBYTES,
+  isCalldataEnveloped,
+  X25519DeoxysII,
+} from '@oasisprotocol/sapphire-paratime';
+import { hexlify, parseUnits, randomBytes } from 'ethers';
+import { randomInt } from 'crypto';
+
+describe('CalldataEncryption', () => {
+  let contract: TestCalldataEncryption;
+
+  before(async () => {
+    const cdeFac = await ethers.getContractFactory('CalldataEncryption');
+    const cdeLib = await cdeFac.deploy();
+    await cdeLib.waitForDeployment();
+
+    const factory = await ethers.getContractFactory('TestCalldataEncryption', {
+      libraries: {
+        CalldataEncryption: await cdeLib.getAddress(),
+      },
+    });
+    contract = await factory.deploy();
+    await contract.waitForDeployment();
+  });
+
+  // Ensures that the JS library provides the same results as Solidity
+  it('testEncryptCallData', async () => {
+    for (let i = 1; i < 1024; i += 1 + i / 5) {
+      const peerKeypair = boxKeyPairFromSecretKey(
+        randomBytes(crypto_box_SECRETKEYBYTES),
+      );
+      const myKeypair = boxKeyPairFromSecretKey(
+        randomBytes(crypto_box_SECRETKEYBYTES),
+      );
+      const calldata = randomBytes(i);
+      const epoch = randomInt(1 << 32);
+      const nonce = randomBytes(15);
```

Comment:
Surely nonce length should be pulled from somewhere else? Instead of defined as `15`

Reply:
Yeah, this could also be slightly surprising?

Reply:
Nonce length exported by the deoxysii package, but not by the Cipher object.

I think just leave this for now as it's in tests and if anything changes it'll break and be easy to find.